File: 0006/0006_zigzag_conversion.go
package main

import "fmt"

const (
	UP   = 0
	DOWN = 1
)

func convert(s string, numRows int) string {
	if numRows <= 1 {
		return s
	}

	b := []byte(s)
	var matrix [][]byte

	for i := 0; i < numRows; i++ {
		matrix = append(matrix, make([]byte, (len(b)+1)/2))
	}

	// start point
	x := 0
	y := 0
	i := 0
	state := DOWN

	for i < len(b) {
		matrix[x][y] = b[i]

		i++
		// find next
		if state == DOWN {
			if x >= numRows-1 {
				x--
				y++
				state = UP
			} else {
				x++
			}
		} else { // state == UP
			if x <= 0 {
				x++
				state = DOWN
			} else {
				x--
				y++
			}
		}
	}

	var r []byte
	for i := 0; i < len(matrix); i++ {
		for j := 0; j < len(matrix[i]); j++ {
			if matrix[i][j] != 0 {
				r = append(r, matrix[i][j])
			}
		}
	}
	return string(r)
}

func assert(expect string, result string) {
	if result != expect {
		panic(fmt.Sprintf("Assert failed!, Expect %s, Get %s", expect, result))
	}
}
func main() {
	assert("PAYPALISHIRING", convert("PAYPALISHIRING", 1))
	assert("PYAIHRNAPLSIIG", convert("PAYPALISHIRING", 2))
	assert("PAHNAPLSIIGYIR", convert("PAYPALISHIRING", 3))
	assert("PINALSIGYAHRPI", convert("PAYPALISHIRING", 4))
}

File: 0006/0006_zigzag_conversion.rs
struct Solution;

impl Solution {
    pub fn convert(s: String, num_rows: i32) -> String {
        let mut rows: Vec<Vec<char>> = vec![vec![]; num_rows as usize];
        let mut idx: i32 = 0;
        let mut dir: i32 = 1;
        for c in s.chars() {
            rows[idx as usize].push(c);
            if dir == 1 && idx >= num_rows-1 {
                dir = -1;
            } else if dir == -1 && idx <= 0 {
                dir = 1;
            }
            if num_rows > 1 {
                idx += dir;
            }
        }
        let mut result: Vec<char> = vec![];
        for v in rows {
            result.extend(v.iter());
        }
        result.iter().collect()
    }
}

fn main() {
	assert_eq!("PAYPALISHIRING".to_string(), Solution::convert("PAYPALISHIRING".to_string(), 1));
	assert_eq!("PYAIHRNAPLSIIG".to_string(), Solution::convert("PAYPALISHIRING".to_string(), 2));
	assert_eq!("PAHNAPLSIIGYIR".to_string(), Solution::convert("PAYPALISHIRING".to_string(), 3));
	assert_eq!("PINALSIGYAHRPI".to_string(), Solution::convert("PAYPALISHIRING".to_string(), 4));
}
